### ISSUE TITLE
Match/render issue with count (range) feature

### ISSUE THREAD

**Issue by mr-tz** [mandiant/capa#1161]
Something appears to be broken her for rendering and/or matching, maybe due to the range/count in the rule?

```
capa tests/data/17264e3126a97c319a6a0c61e6da951e.dll_ -vv -t encrypt

encrypt data using RC4 PRGA
namespace  data-manipulation/encryption/rc4
author     moritz.raabe@mandiant.com
scope      function
att&ck     Defense Evasion::Obfuscated Files or Information [T1027]
mbc        Cryptography::Encrypt Data::RC4 [C0027.009], Cryptography::Generate Pseudo-random Sequence::RC4 PRGA [C0021.004]
function @ 0x5FDDC044
  and:
    match: contain loop @ 0x5FDDC044
      or:
        characteristic: loop @ 0x5FDDC044
    or:

```

match raw data via JSON:
```json
"matches": [
                [
                    {
                        "type": "absolute",
                        "value": 1608368196
                    },
                    {
                        "success": true,
                        "node": {
                            "statement": {
                                "type": "and"
                            },
                            "type": "statement"
                        },
                        "children": [
                            {
                                "success": true,
                                "node": {
                                    "feature": {
                                        "type": "match",
                                        "match": "contain loop"
                                    },
                                    "type": "feature"
                                },
                                "children": [
                                    {
                                        "success": true,
                                        "node": {
                                            "statement": {
                                                "type": "or"
                                            },
                                            "type": "statement"
                                        },
                                        "children": [
                                            {
                                                "success": true,
                                                "node": {
                                                    "feature": {
                                                        "type": "characteristic",
                                                        "characteristic": "loop"
                                                    },
                                                    "type": "feature"
                                                },
                                                "children": [],
                                                "locations": [
                                                    {
                                                        "type": "absolute",
                                                        "value": 1608368196
                                                    }
                                                ],
                                                "captures": {}
                                            },
                                            {
                                                "success": false,
                                                "node": {
                                                    "feature": {
                                                        "type": "characteristic",
                                                        "characteristic": "tight loop"
                                                    },
                                                    "type": "feature"
                                                },
                                                "children": [],
                                                "locations": [],
                                                "captures": {}
                                            },
                                            {
                                                "success": false,
                                                "node": {
                                                    "feature": {
                                                        "type": "characteristic",
                                                        "characteristic": "recursive call"
                                                    },
                                                    "type": "feature"
                                                },
                                                "children": [],
                                                "locations": [],
                                                "captures": {}
                                            }
                                        ],
                                        "locations": [],
                                        "captures": {}
                                    }
                                ],
                                "locations": [
                                    {
                                        "type": "absolute",
                                        "value": 1608368196
                                    }
                                ],
                                "captures": {}
                            },
                            {
                                "success": true,
                                "node": {
                                    "statement": {
                                        "type": "range"
                                    },
                                    "type": "statement"
                                },
                                "children": [],
                                "locations": [],
                                "captures": {}
                            },
                            {
                                "success": true,
                                "node": {
                                    "statement": {
                                        "type": "range"
                                    },
                                    "type": "statement"
                                },
                                "children": [],
                                "locations": [],
                                "captures": {}
                            },
                            {
                                "success": true,
                                "node": {
                                    "statement": {
                                        "type": "range"
                                    },
                                    "type": "statement"
                                },
                                "children": [],
                                "locations": [],
                                "captures": {}
                            },
                            {
                                "success": true,
                                "node": {
                                    "statement": {
                                        "type": "or"
                                    },
                                    "type": "statement"
                                },
                                "children": [
                                    {
                                        "success": false,
                                        "node": {
                                            "feature": {
                                                "type": "match",
                                                "match": "calculate modulo 256 via x86 assembly"
                                            },
                                            "type": "feature"
                                        },
                                        "children": [],
                                        "locations": [],
                                        "captures": {}
                                    },
                                    {
                                        "success": true,
                                        "node": {
                                            "statement": {
                                                "type": "range"
                                            },
                                            "type": "statement"
                                        },
                                        "children": [],
                                        "locations": [],
                                        "captures": {}
                                    }
                                ],
                                "locations": [],
                                "captures": {}
                            },
                            {
                                "success": true,
                                "node": {
                                    "statement": {
                                        "type": "optional"
                                    },
                                    "type": "statement"
                                },
                                "children": [
                                    {
                                        "success": false,
                                        "node": {
                                            "statement": {
                                                "type": "or"
                                            },
                                            "type": "statement"
                                        },
                                        "children": [
                                            {
                                                "success": false,
                                                "node": {
                                                    "feature": {
                                                        "type": "number",
                                                        "number": 255
                                                    },
                                                    "type": "feature"
                                                },
                                                "children": [],
                                                "locations": [],
                                                "captures": {}
                                            },
                                            {
                                                "success": false,
                                                "node": {
                                                    "feature": {
                                                        "type": "number",
                                                        "number": 256
                                                    },
                                                    "type": "feature"
                                                },
                                                "children": [],
                                                "locations": [],
                                                "captures": {}
                                            }
                                        ],
                                        "locations": [],
                                        "captures": {}
                                    }
                                ],
                                "locations": [],
                                "captures": {}
                            }
                        ],
                        "locations": [],
                        "captures": {}
                    }
                ]
            ]
        },
```

**Comment by mike-hunhoff** [comment]
Appears this issue stems from here:
https://github.com/mandiant/capa/blob/da0a1e7903462676355676e9be40e867ab071bbc/capa/render/result_document.py#L171-L184
From `Pydantic` [documentation](https://pydantic-docs.helpmanual.io/usage/types/#unions) on `Field` unions:

_"pydantic will attempt to 'match' any of the types defined under Union and will use the first one that matches. In the above example the id of user_03 was defined as a uuid.UUID class (which is defined under the attribute's Union annotation) but as the uuid.UUID can be marshalled into an int it chose to match against the int type and disregarded the other types."_

TLDR; all `statement` `Field`s are marshaled to `OptionalStatement` here:

https://github.com/mandiant/capa/blob/da0a1e7903462676355676e9be40e867ab071bbc/capa/render/result_document.py#L233-L235

Breaking checks completed later including:

https://github.com/mandiant/capa/blob/da0a1e7903462676355676e9be40e867ab071bbc/capa/render/result_document.py#L277-L278

This prevents features including `Range` and `Subscope` from rendering correctly, and may break other rendering or matching.

I'm not overly familiar with `Pydantic` so I'm still researching how this can be fixed. Pydantic supports something called ["tagged unions"](https://pydantic-docs.helpmanual.io/usage/types/#discriminated-unions-aka-tagged-unions) which may be a start.

If anyone else who is more familiar with `Pydantic` has input I would love to hear your thoughts.

**Comment by mike-hunhoff** [comment]
tagged unions appear to be working:

```
encrypt data using RC4 PRGA
namespace  data-manipulation/encryption/rc4
author     moritz.raabe@mandiant.com
scope      function
att&ck     Defense Evasion::Obfuscated Files or Information [T1027]
mbc        Cryptography::Encrypt Data::RC4 [C0027.009], Cryptography::Generate Pseudo-random Sequence::RC4 PRGA [C0021.004]
function @ 0x5FDDC044
  and:
    match: contain loop @ 0x5FDDC044
      or:
        characteristic: loop @ 0x5FDDC044
    count(characteristic(nzxor)): 1 @ 0x5FDDC177
    count(characteristic(calls from)): 4 or fewer @ 0x5FDACBA8, 0x5FDDC6D5, 0x5FDE6C3A
    count(basic block): 4 or more @ 0x5FDDC044, 0x5FDDC05A, 0x5FDDC060, 0x5FDDC065, and 29 more...
    or:
      count(mnemonic(movzx)): 4 or more @ 0x5FDDC0B0, 0x5FDDC0BA, 0x5FDDC0D7, 0x5FDDC134, and 4 more...
```

**Comment by mike-hunhoff** [comment]
Enabling [smart unions](https://pydantic-docs.helpmanual.io/usage/model_config/#smart-union) also appears to fix this bug and requires many less changes to the existing code. However, we do incur a performance hit as Pydantic checks all possible types before attempting to coerce. #1176 implements a fix using smart unions.

**Comment by mike-hunhoff** [comment]
Unfortunately smart unions do not fix this issue when loading from freeze because there isn't enough context (e.g. object type) to use for coercion. Here we would need to rely on `Union` order to help `pydantic` make decisions but from my testing `Pydantic` appears pretty generous e.g. consider the following contrived example:

```python
...
class BasicOneModel(BaseModel):
    name: str = "basic_one"
    description: Optional[str]

class BasicTwoModel(BaseModel):
    name: str = "basic_two"
    description: Optional[str]

class ComplexModel(BaseModel):
    name: str = "complex"
    description: Optional[str]
    number: int

Child = Union[
    BasicOneModel,
    BasicTwoModel,
    ComplexModel
]

class ParentModel(BaseModel):
    name: str = "parent"
    child: Child

    class Config:
        smart_union = True

node = ComplexModel(number = 100)
print(type(node))

parent = ParentModel(child=node)
print(type(parent.child))

print(type(ParentModel.parse_raw('{"child": {"name": "complex", "number": 100}}').child))
```

This code when executed outputs the following:

```
<class '__main__.ComplexModel'>
<class '__main__.ComplexModel'>
<class '__main__.BasicOneModel'>
```

You can see smart unions work when `pydantic` has objects types to compare but fails when object types are not present for the call to `parse_raw`. Shockingly, `parse_raw` types `ParentModel.child` as `BasicOneModel` even though we specify the unique `number` field of `ComplexModel`. The resulting `BasicOneModel` does not have a `number` field and we lose the value `100` altogether.

Changing the `Union` order to:

```python
Child = Union[
    ComplexModel,
    BasicOneModel,
    BasicTwoModel
]
```

produces expected results:

```
<class '__main__.ComplexModel'>
<class '__main__.ComplexModel'>
<class '__main__.ComplexModel'>
```

I assume that because the fields of `BasicOneModel` are either optional or have a default value `pydantic` considers it an acceptable coercion for all other types. This is most likely the cause of the ordering issue we see in capa here: https://github.com/mandiant/capa/blob/b1b15e2eefe8a8c8975a70c9f4e171b405bae921/capa/features/freeze/features.py#L320-L346

Fortunately, it appears that because the other feature models contain a required, unique field `pydantic` is able to correctly type if we place `BasicBlockFeature` last in the `Union`.

Unfortunately, it does not appear that capa's statement models are unique enough for `pydantic` to correctly type them: https://github.com/mandiant/capa/blob/b1b15e2eefe8a8c8975a70c9f4e171b405bae921/capa/render/result_document.py#L131-L168

We can help `pydantic` by placing `Range`, `Some`, and `Subscope` at the top of the `Union` but there is no way from my understanding for `pydantic` to tell the difference between `And`, `Or`,  `Not`, or `Optional`.

Therefore, I believe the only way to fix this problem is to change capa's statement models 😞
